Implement sum aggregation suggestion with placeholder support

miqh opened the pull request:
Hello,

These changes address #13615.

Apart from implementing the fairly trivial sum aggregation suggestion, I've introduced some basic support for placeholder values within queries. The notation is similar to how the snippet systems in VS Code and Atom. For any query suggestions using placeholders, the query input area should automatically cycle to the next placeholder position in queue, allowing users to choose a value and then pressing <kbd>TAB</kbd> to continue to any remaining.

I considered doing some refactoring around giving these query hint and placeholder abstractions typed interfaces but put it off because getting feedback first would prevent wasted time.

With this sum hint, I've arbitrarily picked ten time series as the threshold after which it starts showing up.

davkal commented on public/app/features/explore/Explore.tsx:
```diff
@@ -306,29 +306,37 @@ export class Explore extends React.PureComponent<ExploreProps, ExploreState> {
-      this.setState({ queries: nextQueries }, () => this.onSubmit());
+      this.queryExpressions = nextState.queries.map(q => q.query);
+      // Accepting certain fixes do not result in a well-formed query which should not be submitted
+      this.setState(nextState, !action.preventSubmit ? this.onSubmit : null);
```
Nice!

davkal commented on public/app/features/explore/QueryField.tsx:
```diff
@@ -116,6 +117,7 @@ export interface TypeaheadFieldState {
   typeaheadPrefix: string;
   typeaheadText: string;
   value: Value;
+  placeholderIndices: number[];
```
I don't think the `placeholderIndices` need to be part of the component state since they do no affect anything inside `render()`. You can make them a local component property instead so that updating them will not cause an additional render update. Also, please file them alphabetically.

davkal commented on public/app/features/explore/QueryField.tsx:
```diff
@@ -295,14 +304,28 @@ class QueryField extends React.PureComponent<TypeaheadFieldProps, TypeaheadField
       suggestionText = onWillApplySuggestion(suggestionText, { ...this.state });
     }
 
+    const placeholderInsertIndex = placeholderIndices[0];
```
I think this logic should in its own function, and not in `applyTypeahead` which is reserved for accepting a suggestion. Can you make a special case in the `Tab` key handler?

davkal commented on public/app/features/explore/utils/prometheus.ts:
```diff
@@ -86,3 +86,35 @@ export function parseSelector(query: string, cursorOffset = 1): { labelKeys: any
+ * 2. `sum( offset 1h) by(|)`
+ * 3. `sum(| offset 1h) by (label)`
+ */
+export function parsePlaceholders(query) {
```
Function could use a type signature.

davkal commented on public/app/features/explore/utils/prometheus.ts:
```diff
@@ -86,3 +86,35 @@ export function parseSelector(query: string, cursorOffset = 1): { labelKeys: any
+      // correctly adjusted to account for placeholder values (e.g. `$1`)
+      // being replaced by empty strings
+      index: index - indexOffset,
+      order: +order,
```
Prefer `parseInt` here.

davkal commented on public/app/plugins/datasource/prometheus/query_hints.ts:
```diff
@@ -1,5 +1,10 @@
+/**
+ * Number of time series results needed before starting to suggest sum aggregation hints
+ */
+export const SUM_HINT_THRESHOLD_COUNT = 10;
```
In other places we have 20 as the limit.

davkal reviewed (changes requested):
This is really awesome and works already well with the `sum()` hint.

This needs a rebase however, since the structure of `Explore.tsx` changed quite a bit. A quick summary of the recent refactor to help you with the rebase:

- querying got disentangled from rows, so now each row sends of their own query
- those queries are called transactions
- each transaction has their own results, errors, and hints

I like your idea of consuming hints, perhaps you can make that work in the new structure as well.

I'm really looking forward for this to get merged. Follow-up PRs could implement snippets around functions like `rate()`. Very cool stuff.

miqh committed:
Add sum aggregation query suggestion

Implements rudimentary support for placeholder values inside a string
with the `PlaceholdersBuffer` class. The latter helps the newly added
sum aggregation query suggestion to automatically focus on the label
so users can easily choose from the available typeahead options.

Related: #13615

miqh committed:
Add tests to cover PlaceholdersBuffer and sum hint

Related: #13615

miqh commented:
@davkal, apologies for taking a while to address your feedback—bit of a hectic week at work.

I've tried my best for a take two. I hope there hasn't been more significant changes.

My main revisions include:

- Encapsulating this placeholder functionality into its own class `PlaceholdersBuffer`, which I've stashed alongside the explore components. I noticed that the typeahead suggestions were refactored into a `LanguageProvider` which hangs off a data source, but I didn't think this placeholder functionality qualified as being language-related.
- Having to debug a strange query inconsistency problem (even without the new placeholder support) which affected how multiple queries behaved on screen if some had hints applied.

davkal reviewed (approved):
Excellent work!

Re the query row sync, I have not seen this again since I switched to the `setState(updater)` pattern. I'm on the fence on whether the explicit sync is a good idea, but I'll keep it in for now. Long-term we'll switch to Redux anyway, then all those state changes will be synchronous again.
